[1-2단계 - 방탈출 예약 대기] 레오(김준영) 미션 제출합니다.

kjyyjk opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
https://prolog.techcourse.co.kr/studylogs/4374

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->
안녕하세요 베루스~
이번 미션 리뷰이인 레오입니다. 잘부탁드려요!

이번 미션에서 진행한 커밋 로그입니다.
https://github.com/woowacourse/spring-roomescape-waiting/pull/245/files/12e1c2f5009c5766bdd0c5f99b0b8202610d2f57..456de34942eed921920d2b34894a4dca672e6d01

### 애플리케이션 실행 관련
애플리케이션 실행 시 admin과 user를 등록하게끔 설정해두었습니다. 로그인 정보는 다음과 같습니다.
```
admin
id : admin@gmail.com
password : qwer!

user
id : user@gmail.com
password : qwer!
```

또한 token 관련 민감 정보를 git에 올리지 않게 해두었습니다. 혹시 실행해보신다면 다음과 같이 properties를 추가하셔야합니다!
  

**resources/application-auth.properties**
```
security.jwt.token.secret_key=3C33F72F97777A4EE3F218448AF1DABS
security.jwt.token.expiration_term=3600000
```


### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

kjyyjk committed:
feat: migrate previous mission code

kjyyjk committed:
chore: jpa 기본 의존성 추가 및 gradle 설정

- DDL 이후 스크립트 실행을 위해 spring.jpa.defer-datasource-initialization=true 설정

Co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

kjyyjk committed:
refactor: 엔티티 연관관계 매핑

Co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

kjyyjk committed:
refactor: Dao를 JpaRepository로 대체

- 추가로 의존성 역전을 통해 Service에서 Jpa 의존성을 제거했다.
덕분에 추후 Jpa가 아닌 다른 기술로 변경할 경우 Repository 구현체만 변경하면 된다.
또한 테스트 시 Fake 객체를 사용할 수 있다.

Co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

yeonnhuu committed:
feat: 내 예약 목록 페이지 구현

Co-authored-by: kjyyjk <rlawnsdud920@gmail.com>

yeonnhuu committed:
feat: 내 예약 목록 API 구현

Co-authored-by: kjyyjk <rlawnsdud920@gmail.com>

kjyyjk committed:
refactor: 예약 상태 표시 리팩토링

Co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

kjyyjk committed:
refactor: 예약 생성 로직 리팩토링

Co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

kjyyjk committed:
refactor: 내 예약 조회 응답 json 수정

Co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

yeonnhuu committed:
refactor: 예약 검색 기능 리팩토링

Co-authored-by: kjyyjk <rlawnsdud920@gmail.com>

kjyyjk committed:
refactor: 클래스 final 키워드 제거

Co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

yeonnhuu committed:
refactor: 예약 검색 기능에서 `@Query` 제거하고 Specification 기반 필터링으로 변경

Co-authored-by: kjyyjk <rlawnsdud920@gmail.com>

kjyyjk committed:
test: 예약 조건 조회 MockMvc 통합 테스트

kjyyjk commented on src/main/java/roomescape/reservation/repository/jpa/ReservationRepositoryImpl.java:
```diff
@@ -0,0 +1,67 @@
+        return reservationJpaRepository.findAllByMemberId(id);
+    }
+
+    @Override
+    public List<Reservation> findAllByCondition(final Long memberId, final Long themeId, final LocalDate from,
+                                                final LocalDate to) {
+        Specification<Reservation> spec = Specification.where(ReservationSpecs.isMemberReservation(memberId))
+                .and(ReservationSpecs.isThemeReservation(themeId))
+                .and(ReservationSpecs.isReservationByPeriod(from, to));
+        return reservationJpaRepository.findAll(spec);
+    }
```
JPQL로 `WHERE (:memberId IS NULL OR :memberId = r.member.id)` 와 같은 쿼리를 작성해 조건부 where 절을 풀어낼 수도 있었지만, 학습 측면에서 JPA가 지원하는 Specification을 사용해봤습니다!

kjyyjk commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -24,94 +23,86 @@ public class ReservationService {
-        return reservationDao.findAll().stream()
-                .filter(totalPredicate)
-                .map(ReservationInfo::new)
-                .toList();
```
기존에는 이러한 검색 조건 필터링도 하나의 비즈니스 로직이라고 보았기 때문에 service에 위치했었습니다.

물론 findAll()로 모든 예약을 조회하는 로직이기에 비효율적이라는 것은 인지했었습니다. 하지만 처음부터 쿼리에 다 녹여내서 복잡한 쿼리를 만드는 것보다는 비즈니스 로직을 우선으로 작성한 후에 성능을 고려하여 하나씩 쿼리에 녹여내는 것이 좋다고 생각했었습니다.

이번에 페어를 진행하면서는 페어와 논의 끝에 전부 Repository의 책임으로 위임했고 현재 Service에서는 제가 생각하는 비즈니스 로직이 드러나지 않고 있습니다.

혹시 베루스도 이러한 필터링 조건들은 비즈니스 로직이라고 보시나요?
성능을 고려한다면 쿼리에 전부 녹여내는 것이 좋은 것인지.. 궁급합니다.

verus-j replied:
상황에 따라 다를 수 있지만, 필터링 조건들도 비즈니스 로직으로 볼 수 있을 것 같습니다. 성능을 고려했다는게 조인문으로 쿼리를 작성했다는걸까요?

kjyyjk replied:
해당 코드 부분은 이전 코드이며, db에서 모든 예약을 조회해온뒤 필터링을 해주고 있습니다.

이에 대해 모든 예약을 조회해오는 것은 비효율적일 수 있다는 의견도 있었습니다.
따라서 페어 프로그래밍을 진행하며 성능도 고려하기 위해 필터링 조건들을 쿼리에 담았습니다.(Specification 활용한 쿼리)

verus-j replied:
아 하나의 테이블에서 조회할 때 여러 조건문들을 조합하는 문제군요. findAll이 어려우면 조건으로 항상 사용해야하는 키(memberId 등)로 우선 조회하여 데이터 수를 줄인 후에 애플리케이션에서 처리할 수 있을 것 같습니다. 물론 이것도 findAll과 같이 데이터가 많으면 지금처럼 쿼리에 필터링 조건을 모두 담을수밖에 없을 것 같네요.

kjyyjk replied:
findAll()로 조회하는 데이터가 많으면 키 같은 조건들을 우선으로 쿼리에 담을 수 있고,
그래도 데이터가 많다면 모든 조건을 쿼리에 담을 수 있겠군요!
감사합니다.

kjyyjk replied:
만약 최적화에 따라 검색 조건을 애플리케이션이 아닌 쿼리에 담도록 변경한다면.
기존에 존재하던 애플리케이션의 조건 필터링 로직을 제거하는게 적절하다고 보시나요?

저는 해당 조건들을 쿼리에 담아냈어도 여전히 비즈니스 로직의 일부라고 보기 때문에ㅡ
애플리케이션에 조건을 확인하는 로직들을 그대로 두는 것이 더 낫다고 생각합니다.

물론 이미 쿼리로 해당 조건들에 맞는 데이터만 조회해와 의미는 없을테지만,
비즈니스 로직을 명시적으로 보인다는 의미가 있다고 생각합니다.

kjyyjk reviewed (commented):
리뷰 잘부탁드립니다~

추가로 이번 JPA를 도입하고나니 대부분의 경우에 쿼리를 직접 작성하지 않아도 돼서 편리했지만,
한편으로는 Repository의 메서드 시그니처에 엔티티의 필드명이 그대로 드러나야하는 것이 불편했습니다.

이 경우 엔티티의 필드명을 수정하면 모든 메서드명도 바꿔야합니다.
만약 메서드명을 변경하지 않으면 컴파일 시점에는 아무 에러도 발견되지 않다가 런타임 시점에 에러가 발생합니다.

이러한 위험?이 존재하는데 과연 JPA가 마냥 좋은 것인지는 아직까지 잘 모르겠습니다.

verus-j commented on src/main/java/roomescape/reservation/repository/jpa/ReservationRepositoryImpl.java:
```diff
@@ -0,0 +1,67 @@
+import roomescape.reservation.repository.ReservationRepository;
+
+@Repository
+public class ReservationRepositoryImpl implements ReservationRepository {
```
대부분 ReservationJpaRepository로 위임하는 메서드들인데 굳이 별도로 클래스를 분리한 이유가 있을까요?

kjyyjk replied:
크게 두 가지 이유가 존재합니다.

첫 번째는 service가 특정 기술에 강결합되지 않도록 하기 위함입니다.
jdbc -> jpa로 전환하는 과정에서 repository의 인터페이스를 변경해야만 했었고, 이로 인해 service에도 변경이 전파되는 것을 경험했습니다. 이는 service가 특정 db 구현체에 의존해있었기 때문이라고 판단했습니다. 따라서 dip 개념을 활용해 의존성을 관리해주었습니다. 물론 당장의 미션에서 또 다른 변경이 발생하지는 않겠지만, 이는 오버 엔지니어링이 아니라고 생각하며 단점보다는 장점이 더 크다고 생각합니다.

두 번째는 테스트 편리성에 있습니다.
Service를 단위 테스트하는 경우 FakeRepository와 JUnit을 사용할 수 있습니다. 만약 Service가 JpaRepository에 직접 의존하고 있다면 FakeRepository가 Jpa 인터페이스를 전부 구현해야하는 문제가 발생합니다. 결국 첫 번째 이유와 같은 맥락으로, 별도로 클래스를 분리하면 테스트 시 FakeRepository 구현체로 갈아끼울 수 있다는 장점이 있습니다.

verus-j replied:
1. JPA가 DB의 구현체를 감싸주지 않나요? JpaRepository를 상속한 ReservationRepository 인터페이스를 활용하면 다른 기술(jdbc 등)을 사용해도 인터페이스의 구현체만 작성하면 되니 Service의 변경은 없지 않을까요?

2. Jpa 인터페이스를 전부 구현하지 않고 사용하는 방법이 있습니다. 필요한 메서드를 정의한 인터페이스를 만들고 그걸 JpaRepository와 함께 상속시키면 됩니다. 그러면 따로 위임 메서드를 구현할 필요없이 인터페이스만 가지고 사용할 수 있습니다. 자세한건 한번 GPT에게 물어봐주세요~

kjyyjk replied:
말씀주신 1번에 대해서는 동의하고 설득됐습니다!
사실 DB 구현체가 아니라 DB 기술을 말씀드리고 싶었는데, 다른 DB 기술을 도입하더라도 충분히 ReservationRepository를 
활용할 수 있을 것 같아요.

하지만 2번에 대해서는 잘 이해하지 못했습니다. 제가 이해한 구조는 아래 그림과 같은데요..!
CustomRepository에 필요한 메서드를 정의하고 이를 Fake가 구현했을 때는 Fake가 Repository를 대체하지는 못할 것 같습니다. 그렇다고 Fake가 Repository를 구현한다면, JpaRepository의 메서드를 전부 구현해야하는 문제가 여전히 발생할 것 같습니다.

이에 대해 gpt와 몇몇 크루와도 얘기했지만 해결하지 못했는데요, 혹시 조금 더 방향성을 제시해주실 수 있을까요?😢

<img width="848" alt="image" src="https://github.com/user-attachments/assets/89bed875-1955-4580-9837-2e2f1ec3d2bc" />

verus-j replied:
<img width="848" alt="445517547-89bed875-1955-4580-9837-2e2f1ec3d2bc" src="https://github.com/user-attachments/assets/5e522f0c-9963-4820-b68b-5c1ce4f0b120" />


```java
@Repository
interface MemberJpaRepository extends JpaRepository<Member, Long>, MemberRepository { }

public interface MemberRepository {

    Optional<Member> findByEmailAndPassword(String email, String password);

    Optional<Member> findById(long memberId);

    Member save(Member member);

    List<Member> findAll();
}
```

이렇게 구성하면 Jpa 구현체는 MemberJpaRepository 타입으로 생성하되 public 접근자가 아니여서 외부에서 직접적으로 접근을 못하고 MemberRepository를 통해서만 접근 가능합니다. 여기서 Fake 객체를 만들 때 MemberRepository 타입으로 만들면 프로덕션 코드와 테스트 코드에서 모두 MemberRepository 타입으로 Repository를 사용할 수 있습니다. Jpa 구현체는 MemberRepository의 하위 타입인 MemberJpaRepository의 하위 타입을 생성할테니 당연히 MemberRepository로 추상화해서 사용할 수 있겠죠!

결국 서비스에서는 DB 접근과 관련된 기술을 MemberRepository 인터페이스 내부에 숨겨졌기 때문에 관심사가 분리되겠네요~

verus-j commented on src/main/java/roomescape/reservation/repository/jpa/ReservationTimeRepositoryImpl.java:
```diff
@@ -0,0 +1,43 @@
+import roomescape.reservation.repository.ReservationTimeRepository;
+
+@Repository
+public class ReservationTimeRepositoryImpl implements ReservationTimeRepository {
```
여기는 모든 메서드가 단순 위임 밖에 없어서 외부에서 바로 ReservationTimeJpaRepository를 사용해도 되지 않을까요?

kjyyjk replied:
위 코멘트와 같은 이유입니다!
https://github.com/woowacourse/spring-roomescape-waiting/pull/245#discussion_r2094107156

verus-j replied:
https://github.com/woowacourse/spring-roomescape-waiting/pull/245#discussion_r2094753085

verus-j commented on src/main/java/roomescape/reservation/repository/jpa/ThemeRepositoryImpl.java:
```diff
@@ -0,0 +1,49 @@
+import roomescape.reservation.repository.ThemeRepository;
+
+@Repository
+public class ThemeRepositoryImpl implements ThemeRepository {
```
여기도 별도로 분리한 이유가 궁금합니다.

kjyyjk replied:
위 코멘트와 같은 이유입니다!
https://github.com/woowacourse/spring-roomescape-waiting/pull/245#discussion_r2094107156

verus-j replied:
https://github.com/woowacourse/spring-roomescape-waiting/pull/245#discussion_r2094753085

verus-j commented on src/test/java/roomescape/reservation/service/ReservationServiceTest.java:
```diff
@@ -181,15 +188,15 @@ void findAll() {
         // when
         // then
-        assertThat(reservationService.getReservations()).hasSize(2);
+//        assertThat(reservationService.getReservations()).hasSize(2);
```
여기는 주석 처리한 이유가 있나요?

kjyyjk replied:
전체적으로 불필요한 주석 제거하여 다시 리뷰 요청 드리겠습니다!

verus-j commented on src/test/java/roomescape/reservation/service/ReservationServiceTest.java:
```diff
@@ -201,7 +208,71 @@ void cancelById() {
+
+    @DisplayName("멤버 예약 목록을 조회할 수 있다")
+    @Test
+    void findReservationsByMember() {
```
테스트가 너무 복잡한 것 같습니다. 테스트 코드의 가독성을 챙길 수 없을까요?

kjyyjk replied:
이에 대해 인지하고 있었지만 쉽게 손대지 못하고 있었습니다.
이번 기회에 전체적으로 리팩터링하고 가독성을 높여보도록 하겠습니다!

verus-j commented on build.gradle:
```diff
@@ -21,7 +21,7 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-validation'
```
테스트가 실패하는게 많네요 ㅠㅠ
확인부탁드려요~

kjyyjk replied:
앗 분명 테스트 다 통과하는거 확인하고 리뷰 요청드렸었는데, 생각해보니 auth 관련 설정 안내를 안드렸었네요ㅠ
죄송합니다.

kjyyjk replied:
token 관련 민감 정보를 git에 올리지 않게 해두었습니다. 다음과 같이 properties를 추가하셔야합니다..!
(추가적인 내용은 pr 본문에 추가해두었습니다.)

**resources/application-auth.properties**
```
security.jwt.token.secret_key=3C33F72F97777A4EE3F218448AF1DABS
security.jwt.token.expiration_term=3600000
```

verus-j reviewed (changes requested):
안녕하세요 레오
미션 진행하느라 고생많으셨습니다.
피드백 남겨뒀으니 확인부탁드려요~

kjyyjk committed:
refactor: 단위 테스트 데이터 셋업을 위한 BeforeEach 메서드 도입

- 테스트 중복 코드 제거 및 가독성 향상을 위해 BeforeEach 메서드를 추가하여 공통 테스트 데이터를 초기화한다.

kjyyjk committed:
refactor: TokenPayload 및 AdminRoleInterceptor 검증 로직 리팩터링

- TokenInfo DTO로 TokenPayload를 관리하도록 수정했다.

- Role까지는 민감한 정보가 아니라고 판단했기에 JwtProvider에서 토큰 생성 시
회원 ID와 Role을 포함하도록 수정했다. 토큰이 Role을 포함하고 있으면
어드민 권한 인터셉터에서 쿼리 없이 사용자 인가 처리를 할 수 있다.

kjyyjk committed:
test: @SpringBootTest → @DataJpaTest로 변경

- 해당 테스트들은 web 관련 컴포넌트들을 빈으로 등록할 필요가 없기 때문에,
@DataJpaTest에 필요한 빈들만 추가로 임포트해주었다.

kjyyjk commented:
@coderabbitai resources에 있는 프론트엔드 파일은 리뷰 안 해줘도 돼!

kjyyjk committed:
refactor: 테스트 환경 리팩터링

- application-test.properties 파일 추가로 테스트용 설정 분리해서
테스트 시에는 @Sql로 지정한 스크립트 외 자동 스크립트 실행을 막는다.

- member-data.sql을 추가했다.

- @ActiveProfiles 어노테이션 추가로 테스트 환경에 따른 프로필(test, auth) 명시했다.

- 하나의 테스트 메서드마다 스프링 컨텍스트를 다시 띄워야하지 않는 경우일때는
다른 테스트 클래스와의 독립성을 보장하기 위해 DirtiesContext를 BEFORE_CLASS로 지정했다.

kjyyjk committed:
test: 통합 테스트 클래스에 @Transactional 추가

- DirtiesContext로 매 테스트 메서드마다 컨텍스트를 새로 띄우는 경우에는
필요가 없어 사용하지 않았다.

verus-j commented on src/test/java/roomescape/integration/TimeServiceIntegrationTest.java:
```diff
@@ -32,6 +33,7 @@
 @DataJpaTest
 @Import(value = {ReservationTimeRepositoryImpl.class, ReservationRepositoryImpl.class})
 @DirtiesContext(classMode = ClassMode.BEFORE_CLASS)
+@Transactional
```
이미 `@Transactional`로 DB에 저장되지 않고 롤백 될텐데 `@DirtiesContext`를 유지한 이유가 있을까요?

kjyyjk replied:
말씀주신 것처럼 `@Transacional`이 있기 때문에 각 테스트마다 롤백되는 것을 보장받습니다.
하지만 명시적으로 그리고 확실하게 각 테스트 클래스들이 독립적으로 실행되는 것을 보장하기 위해 `@DirtiesContext(classMode = ClassMode.BEFORE_CLASS)`를 유지했습니다.

이는 테스트 클래스 실행 이전에 한번만 스프링 컨텍스트를 새로 로드하기 때문에 그 비용이 크지 않습니다.

정리하자면 `@Transacional`을 사용한 것은 테스트 메서드 간 db 격리를 보장하기 위한 것이고,
`@DiriesContext`는 테스트 클래스 간 컨텍스트 격리를 보장하기 위한 것입니다.

그리고 이 두 어노테이션을 적절한 속성과 함께 사용한다면 잘 격리된 테스트 환경을 구성할 수 있다고 생각합니다.

verus-j replied:
빈으로 등록된 객체들은 대부분 상태가 변경되지 않는 객체 아닌가요? 컨텍스트 격리가 필요하다면 `@DirtiesContext`를 사용하는게 맞지만, 지금 구현에서는 필요한지 의문입니다.

verus-j commented on src/main/java/roomescape/reservation/controller/ReservationController.java:
```diff
@@ -0,0 +1,80 @@
+        return ResponseEntity.ok().body(responses);
+    }
+
+    @GetMapping("/reservations-mine")
+    public ResponseEntity<List<MemberReservationResponse>> findAllMine(LoginMemberInfo loginMemberInfo) {
+        final List<ReservationInfo> reservationInfos = reservationService.findReservationsByMemberId(loginMemberInfo.id());
+        final List<MemberReservationResponse> responses = reservationInfos.stream()
+                .map(MemberReservationResponse::new)
+                .toList();
+        return ResponseEntity.ok().body(responses);
+    }
```
이번에 추가된 기능인데 컨트롤러의 테스트가 없네요~

kjyyjk replied:
테스트 추가 했습니다.

verus-j commented on src/main/java/roomescape/reservation/repository/jpa/ReservationRepositoryImpl.java:
```diff
@@ -0,0 +1,67 @@
+        return reservationJpaRepository.findAllByDateAndThemeId(date, themeId);
+    }
+
+    @Override
+    public List<Reservation> findAllByMemberId(long id) {
+        return reservationJpaRepository.findAllByMemberId(id);
+    }
```
여기 테스트가 빠졌습니다.

kjyyjk replied:
테스트 추가 했습니다.

verus-j reviewed (changes requested):
주말동안 일정이 있어 리뷰가 늦었습니다.
몇가지 피드백 남겨뒀으니 확인부탁드려요~

kjyyjk committed:
test: 어드민 예약 생성 및 본인 예약 조회 테스트 추가

verus-j reviewed (approved):
안녕하세요 레오
피드백 반영하시느라 고생하셨습니다.
다음 단계 진행해주세요~
